From: Jan Kara <jack@suse.cz>
Subject: quota: check presence of quota operation structures instead of ->quota_read and ->quota_write callbacks
Date: Thu, 13 Jul 2023 15:48:45 +0200

Currently we check whether superblock has ->quota_read and ->quota_write
operations to check whether filesystem supports quotas.  However for
example for shmfs we will not read or write dquots so check whether quota
operations are set in the superblock instead.

Link: https://lkml.kernel.org/r/20230713134848.249779-4-cem@kernel.org
Signed-off-by: Jan Kara <jack@suse.cz>
Reviewed-by: Carlos Maiolino <cmaiolino@redhat.com>
Cc: Al Viro <viro@zeniv.linux.org.uk>
Cc: Christian Brauner <brauner@kernel.org>
Cc: Darrick J. Wong <djwong@kernel.org>
Cc: Hugh Dickins <hughd@google.com>
Cc: Lukas Czerner <lczerner@redhat.com>
Signed-off-by: Andrew Morton <akpm@linux-foundation.org>
---

 fs/quota/dquot.c |    2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

--- a/fs/quota/dquot.c~quota-check-presence-of-quota-operation-structures-instead-of-quota_read-and-quota_write-callbacks
+++ a/fs/quota/dquot.c
@@ -2367,7 +2367,7 @@ int dquot_load_quota_sb(struct super_blo
 
 	if (!fmt)
 		return -ESRCH;
-	if (!sb->s_op->quota_write || !sb->s_op->quota_read ||
+	if (!sb->dq_op || !sb->s_qcop ||
 	    (type == PRJQUOTA && sb->dq_op->get_projid == NULL)) {
 		error = -EINVAL;
 		goto out_fmt;
_
